lib/lookup/src/lookup_buf/mod.rs

```diff
@@ -0,0 +1,399 @@
+            segments: segments.into_iter().collect(),
+        }
+    }
+
+    /// Return a borrow of the SegmentBuf set.
+    #[instrument(level = "trace")]
+    pub fn as_segments(&self) -> &VecDeque<SegmentBuf> {
+        &self.segments
+    }
+
+    /// Return the SegmentBuf set.
+    #[instrument(level = "trace")]
+    pub fn into_segments(self) -> VecDeque<SegmentBuf> {
+        self.segments
+    }
+
+    /// Merge a lookup.
+    #[instrument(level = "trace")]
+    pub fn extend(&mut self, other: Self) {
+        self.segments.extend(other.segments)
+    }
+
+    /// Returns `true` if `needle` is a prefix of the lookup.
+    #[instrument(level = "trace")]
+    pub fn starts_with(&self, needle: &LookupBuf) -> bool {
+        needle.iter().zip(&self.segments).all(|(n, s)| n == s)
+    }
+
+    /// Create the possible fields that can be followed by this lookup.
+    /// Because of coalesced paths there can be a number of different combinations.
+    /// There is the potential for this function to create a vast number of different
+    /// combinations if there are multiple coalesced segments in a path.
+    #[instrument(level = "trace")]
+    pub fn to_alternative_components(&self) -> Vec<Vec<&str>> {
+        let mut components = vec![vec![]];
+        for segment in &self.segments {
+            match segment {
+                SegmentBuf::Field(FieldBuf { name, .. }) => {
+                    for component in &mut components {
+                        component.push(name.as_str());
+                    }
+                }
+
+                SegmentBuf::Coalesce(fields) => {
+                    components = components
+                        .iter()
+                        .flat_map(|path| {
+                            fields.iter().map(move |field| {
+                                let mut path = path.clone();
+                                path.push(field.name.as_str());
+                                path
+                            })
+                        })
+                        .collect();
+                }
+
+                SegmentBuf::Index(_) => {
+                    return Vec::new();
+                }
+            }
+        }
+
+        components
+    }
+}
+
+impl FromStr for LookupBuf {
+    type Err = LookupError;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let lookup = Lookup::from_str(input)?;
+        let lookup_buf: LookupBuf = lookup.into();
+        Ok(lookup_buf)
+    }
+}
+
+impl IntoIterator for LookupBuf {
+    type Item = SegmentBuf;
+    type IntoIter = std::collections::vec_deque::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.segments.into_iter()
+    }
+}
+
+impl From<String> for LookupBuf {
+    fn from(input: String) -> Self {
+        let mut segments = VecDeque::with_capacity(1);
+        segments.push_back(SegmentBuf::from(input));
+        LookupBuf { segments }
+        // We know this must be at least one segment.
```

Comment:
I'm not really sure what this comment is implying here (and in the other `From`s as well)?

Reply:
Also, shouldn't this be `TryFrom`, since the string might not be a valid path?

Reply:
Ah, those comments hark back to when lookups had to have at least one segment. The implication being that since it can guarantee there is one segment in this Lookup, it doesn't need to call `lookup.is_valid()`. I'll remove the comments.

These functions take the whole string as a single segment, there is no parsing here so cannot fail. So `LookupBuf::from("hello.zork[2]"` is the same as the path: `."hello.zork[2]"`.